src/UniformDocs/Helpers/ServiceWorkerMiddleWare.cs

```
@@ -0,0 +1,158 @@
+        /// using the template fetched from the static file server.
+        /// </summary>
+        public ServiceWorkerMiddleWare()
+        {
+           
+        }
+
+        /// <summary>
+        /// Creates a new instance of <see cref="ServiceWorkerMiddleWare"/>
+        /// using the given standalone page template.
+        /// </summary>
+        public ServiceWorkerMiddleWare(string standaloneTemplate)
+        {
+            if (string.IsNullOrEmpty(standaloneTemplate)) throw new ArgumentNullException("standaloneTemplate");
+            appShell = standaloneTemplate;
+        }
+
+        void IMiddleware.Register(Application application)
+        {
+            application.Use(request =>
+            {
+                if (request.Uri.Equals("/sw.js"))
+                {
+                    string template;
+                    if (!string.IsNullOrEmpty(serviceWorkerTemplate))
+                    {
+                        template = serviceWorkerTemplate;
+                    }
+                    else
+                    {
+                        template = FetchServiceWorkerTemplate();
+                        serviceWorkerTemplate = template;
+                    }
+                    var SW = template.Replace("REPLACE_ME_WTH_RUNTIME_HASH", RUNTIME_CACHE_KEY);
+                    var response = new Response
+                    {
+                        BodyBytes = defaultEncoding.GetBytes(SW)
+                    };
+                    response.Headers["Content-Type"] = "text/javascript";
+                    request.SendResponse(response);
+                }
+                return null;
+            });
+
+            application.Use(MimeProvider.Html(this.Invoke));
+        }
+
+        private static string FetchAppShellTemplate()
+        {
+            try
+            {
+                string appShellHTMLUrl = "/sys/app-shell/app-shell-with-service-worker.html";
+                return Self.GET(appShellHTMLUrl).Body;
```

The app shell with a service worker should be fetched only if the original URL contains `https://`. Is there a way to know that?

---

But it works in localhost without HTTPS.

---

Ok so also check for the hostname?

---

I'll find a way. Maybe it fails silently in HTTP which would be cool.

---

Yup. It fails silently in Chrome and console logs this in FF "Error during service worker registration: DOMException: "The operation is insecure.". Do we need to bother?

---

Maybe not for now.